### PR TITLE
CNN Support Added

### PR DESCRIPTION
**Pull request by Prasanna28Devadiga**
Solves issue  #77 
Model is configured via the YAML file. Have provided an yaml file example. Tested on Colab. Utilising Keras ensured that we could support a variety of layers besides just the standard Conv2D. Have added docstrings as per the contributing guidelines
@nidhaloff Is this adequate? Kindly give me a review

### CODE REVIEW

File: dl_support/igel_cnn.py [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+
+
+class CNN:
+  def __init__(self,path):
+    """
+    The YAML file is parsed and all training parameters/ dataset arguments are set. 
+    We stick with the defaults in case the user doesnt provide them
+
+    """
+    with open(path, 'rb') as f:
```
[the remark this answers is not in the document]
Comment (Prasanna28Devadiga) [reply dl_support/igel_cnn.py:20]: Alright will do

File: dl_support/igel_cnn.py [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+      conf = yaml.safe_load(f.read()) 
+    self.batch_size= conf['model']['arguments'].get('batch_size',32)
```
[the remark this answers is not in the document]
Comment (Prasanna28Devadiga) [reply dl_support/igel_cnn.py:22]: Yes right i should be doing that

File: dl_support/igel_cnn.py [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+      pool_size_2d= i[1].get('parameters').get('pool_size',[2,2])
+      pool_size_3d= i[1].get('parameters').get('pool_size',[2,2,2])
+      stride_2d= tuple(stride_2d)
+      stride_3d=tuple(stride_3d)
+      kernel_size_2d= tuple(kernel_size_2d)
+      kernel_size_3d= tuple(kernel_size_3d)
+      pool_size_2d= tuple(pool_size_2d)
+      pool_size_3d= tuple(pool_size_3d)
+      
+      if x == "Dense":
```
[the remark this answers is not in the document]
Comment (Prasanna28Devadiga) [reply dl_support/igel_cnn.py:66]: Hey thats a pretty good idea. The challenge i was faced with was to map these strings to the respective class layers. I think i should be able to use a dictionary to do that that , and maybe just remove all this clutter from the main file

File: igel.yaml [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+  arguments:
+    batch_size: 32
+    epochs: 20
+    validation_split: 0.2
+    filename: "train.csv"
+    color_mode: "grayscale"
+    class_mode: "sparse"
+    target_size: [28,28]
+    loss: "SparseCategoricalCrossentropy"
+    model_layers:
```
[the remark this answers is not in the document]
Comment (Prasanna28Devadiga) [reply igel.yaml:23]: No particular reason actually. I'll make commits with the example you mentioned if thats required.

File: dl_support/igel_cnn.py [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+        model.add(keras.layers.GlobalAveragePooling2D())
+      elif x == "GlobalAveragePooling3D":
+        model.add(keras.layers.GlobalAveragePooling3D())
+      else:
+        print("Please enter a valid layer") 
+      print(model.summary())
+    #model.summary()
+    return model
+
+  def generate_dataset(self):
```
[the remark this answers is not in the document]
Comment (Prasanna28Devadiga) [reply dl_support/igel_cnn.py:108]: Oh yes its dynamic. All it does is read the data from a data frame named as train.csv , and then create two image data-loaders.
 the first column of the dataframe will be assumed to contain the paths , and the next column the corresponding labels.

Talking about mnist though i just realized the default value for the imagesize variable is (784,). This obviously is a bug as the default value should be (256^2, ) considering the data generator resizes all images to (256,256). I'll fix this with the next commit.

**Comment by Prasanna28Devadiga** [comment]
@nidhaloff Thank you for the detailed review. This was extremely insightful. As a result I've noticed a few key areas where improvements can be made , and some bug fixes as well.
I intend to fix these as soon as possible. I'll even add a few unit tests with the next commit. My university exams are due so might take a bit longer. Let me know if that's fine :)

**Comment by HeerakKashyap** [comment]
Does the current implementation support datasets other than MNIST out of the box? If not, maybe a note in the documentation or a TODO comment could help guide future improvements.
